fedimint-client/src/client_rpc.rs

```diff
@@ -0,0 +1,159 @@
+use std::marker::PhantomData;
+use std::{collections::BTreeMap, pin::Pin};
+
+use anyhow::Context;
+use fedimint_core::task::{MaybeSend, MaybeSync};
+use fedimint_core::{apply, async_trait_maybe_send};
+use futures::{Stream, StreamExt as _};
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct RpcHandlerKey<'a> {
+    // Q: should we use instance id instead of module kind?
+    module: &'a str,
+    method: &'a str,
+}
+pub struct Server<Ctx> {
+    handlers: BTreeMap<RpcHandlerKey<'static>, Box<dyn DynClientRpcHandler<Ctx>>>,
+}
+
+impl<Ctx> Server<Ctx> {
+    pub fn new() -> Self {
+        Server {
+            handlers: BTreeMap::new(),
+        }
+    }
+
+    pub fn add_handler<HCtx, H: ClientRpcSingleHandler<HCtx>>(&mut self, handler: H)
+    where
+        HCtx: MaybeSend + MaybeSync + 'static,
+        Ctx: MaybeSend + MaybeSync + AsRef<HCtx>,
+        H: MaybeSend + MaybeSync + 'static,
+    {
+        let handler = Box::new(DynClientRpcHandlerWrapper(handler, PhantomData));
+        self.handlers.insert(
+            RpcHandlerKey {
+                module: H::MODULE,
+                method: H::METHOD,
+            },
+            handler,
+        );
+    }
+
+    pub fn handle_request<'a>(
+        &'a self,
+        ctx: &'a Ctx,
+        module: &'a str,
+        method: &'a str,
+        body: serde_json::Value,
+    ) -> impl Stream<Item = anyhow::Result<serde_json::Value>> + 'a {
+        if let Some(handler) = self.handlers.get(&RpcHandlerKey { module, method }) {
+            handler.handle(ctx, body)
+        } else {
+            Box::pin(futures::stream::once(async {
+                Err(anyhow::anyhow!("Handler not found"))
+            }))
+        }
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+pub trait ClientRpcHandler<Ctx> {
```

Comment:
Do we need another trait? Couldn't this be a method on the existing `ClientModule`, just like cli is on it already?